### PR TITLE
*: http and https on the same port

### PR DESCRIPTION
**Pull request by xiang90**
```
./etcd --experimental-v3demo --cert-file cert --key-file key \
--listen-client-urls=http://127.0.0.1:2379,https://127.0.0.1:2379 \
--advertise-client-urls=http://127.0.0.1:2379,https://127.0.0.1:2379
```

works now.

fix https://github.com/coreos/etcd/issues/4737

/cc @heyitsanthony @gyuho @philips

### CODE REVIEW

**Comment by gyuho** [comment]
Test failed from govet. I have quick fix here https://github.com/coreos/etcd/pull/4833.

Could you merge that first, and rerun this test?

Thanks.

File: etcdmain/etcd.go [hunk 1]
```diff
@@ -204,30 +203,25 @@ func startEtcd(cfg *config) (<-chan struct{}, error) {
...
-
 	if !cfg.peerTLSInfo.Empty() {
 		plog.Infof("peerTLS: %s", cfg.peerTLSInfo)
 	}
-
 	plns := make([]net.Listener, 0)
 	for _, u := range cfg.lpurls {
 		if u.Scheme == "http" && !cfg.peerTLSInfo.Empty() {
 			plog.Warningf("The scheme of peer url %s is http while peer key/cert files are presented. Ignored peer key/cert files.", u.String())
 		}
```
Comment (gyuho) [inline etcdmain/etcd.go]: Deleting `peerAutoTLS`? Then where do we handle auto-TLS?

File: etcdmain/etcd.go [hunk 2]
```diff
@@ -316,40 +332,25 @@ func startEtcd(cfg *config) (<-chan struct{}, error) {
...
-	ch := &cors.CORSHandler{
+	ch := http.Handler(&cors.CORSHandler{
 		Handler: etcdhttp.NewClientHandler(s, srvcfg.ReqTimeout()),
 		Info:    cfg.corsInfo,
-	}
+	})
 	ph := etcdhttp.NewPeerHandler(s)
 
-	var grpcS *grpc.Server
-	if cfg.v3demo {
```
Comment (gyuho) [inline etcdmain/etcd.go]: So we enable v3 by default after this patch?

File: etcdmain/serve.go [hunk 3]
```diff
@@ -15,37 +15,87 @@
...
-		go plog.Fatal(grpcS.Serve(grpcl))
+func serve(sctx *serveCtx, s *etcdserver.EtcdServer, tlscfg *tls.Config, handler http.Handler) error {
+	logger := defaultLog.New(ioutil.Discard, "etcdhttp", 0)
+
+	m := cmux.New(sctx.l)
+
+	if sctx.insecure {
+		gs := v3rpc.Server(s, nil)
+		grpcl := m.Match(cmux.HTTP2())
+		go func() { go plog.Fatal(gs.Serve(grpcl)) }()
```
Comment (gyuho) [inline etcdmain/serve.go]: Why another goroutine? Do we mean `go func() { plog.Fatal(gs.Serve(grpcl)) }()`?

File: etcdmain/etcd.go (hunk 1 above)
Comment (xiang90) [inline etcdmain/etcd.go]: i do not understand this.

File: etcdmain/etcd.go (hunk 2 above)
Comment (xiang90) [inline etcdmain/etcd.go]: sort of. we still need to remove this flag in a few other places.

File: etcdmain/etcd.go (hunk 1 above)
Comment (xiang90) [inline etcdmain/etcd.go]: ok. added it back.

Comment (gyuho) [inline etcdmain/etcd.go]: thanks!

File: etcdmain/etcd.go (hunk 2 above)
Comment (gyuho) [inline etcdmain/etcd.go]: ok got it.

**Comment by xiang90** [comment]
I still need to fix a few tests.

**Comment by xiang90** [comment]
@heyitsanthony @gyuho Now test passed. PTAL.

**Comment by heyitsanthony** [comment]
code looks fine (although `startEtcd` really needs to be split up), but can this include a small e2e test?

**Comment by xiang90** [comment]
@heyitsanthony Will add a e2e test.

**Comment by xiang90** [comment]
@heyitsanthony Added an e2e test.

File: e2e/etcd_test.go [hunk 4]
```diff
@@ -172,14 +210,16 @@ type etcdProcess struct {
...
 
 type etcdProcessClusterConfig struct {
-	clusterSize   int
-	proxySize     int
-	isClientTLS   bool
+	clusterSize int
+	proxySize   int
+	isClientTLS bool
+	// serve both TLS and non-TLS
+	isClientBoth  bool
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: `isClientTLS` and `isClientBoth` should probably be merged into one enum field `clientTLS` with values `None, TLS, Both`

Comment (xiang90) [inline e2e/etcd_test.go]: will do.

**Comment by heyitsanthony** [comment]
lgtm aside from the isClientBoth stuff

**Comment by xiang90** [comment]
@heyitsanthony Fixed. Merging...

**Pull request merged**